### PR TITLE
Add Context Parallelism to Flux model training

### PR DESCRIPTION
**Pull request by limou102**
**1) Add Context Parallelism(CP) support to Flux model training**
Context Parallelism mainly used for video generation models, in the Flux model, the sequence length used in attention computations is very small(512), so context parallelism provides no speedup, other multimodal models can refer to this modification.

The comparison of loss curves with CP enabled/disabled is shown below (gray represents CP=4), with the same global_batch_size=32.
image

The validation loss curve(with coco dataset) is shown below.
image

**2) fix compatibility issues between the Flux code and the latest main branch**

### CODE REVIEW

File: torchtitan/experiments/flux/train.py [hunk 1]
```diff
@@ -90,6 +90,13 @@ def __init__(self, job_config: JobConfig):
     def forward_backward_step(
         self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor
     ) -> torch.Tensor:
+
+        cp_group = (
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: we don't need a handle to `cp_group`. `self.parallel_dims.cp_enabled` is enough to tell if CP is enabled.

File: torchtitan/experiments/flux/train.py [hunk 2]
```diff
@@ -132,24 +139,50 @@ def forward_backward_step(
...
-            pred = unpack_latents(latent_noise_pred, latent_height, latent_width)
+        # limou
+        with torch.no_grad(), torch.device(self.device):
             target = noise - labels
-            loss = self.loss_fn(pred, target)
-        # pred.shape=(bs, seq_len, vocab_size)
-        # need to free to before bwd to avoid peaking memory
-        del (pred, noise, target)
-        loss.backward()
+            target = pack_latents(target)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: ```pred = unpack_latents(latent_noise_pred, latent_height, latent_width)```
to 
```target = pack_latents(target)```
Are they equivalent? Why are we making this change?

File: torchtitan/experiments/flux/train.py [hunk 3]
```diff
@@ -132,24 +139,50 @@ def forward_backward_step(
...
-                img_ids=latent_pos_enc,
-                txt=t5_encodings,
-                txt_ids=text_pos_enc,
-                y=clip_encodings,
-                timesteps=timesteps,
-            )
-
-            # Convert sequence of patches to latent shape
-            pred = unpack_latents(latent_noise_pred, latent_height, latent_width)
+        # limou
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: pls remove this line

File: torchtitan/experiments/flux/train.py (hunk 2 above)
Comment (limou102) [reply torchtitan/experiments/flux/train.py]: These two approaches:
```
pred = forward(input)
pred = unpack_latents(pred)
loss = self.loss_fn(pred, target) # both in unpacked format
```

```
target = pack_latents(target)
pred = forward(input)
loss = self.loss_fn(pred, target) # both in packed format
```
should be functionally equivalent.

The reason is that pack_latents and unpack_latents only perform a reshape operation on tensors, converting image format (batch, channel, height, width) into sequence format (batch, seqlen, hidden_size). The total number of elements remains unchanged, so the loss computation result is unaffected. For example, in the Flux model, the MSELoss depends only on the total number of elements.
image

As for why the code was changed:
One reason is to make it consistent with other models (e.g., LLaMA) where context parallelism (CP) is implemented in the same way.
Another reason is related to how the torchtitan framework handles loss under CP.
When CP is enabled, torchtitan computes loss statistics by averaging across each CP rank.

https://github.com/pytorch/torchtitan/blob/6a3a9da9564d82a1120c7639ef6236bb4cffa049/torchtitan/train.py#L533

**After enabling CP (Context Parallelism), the total number of elements N used in loss computation will be divided by a CP factor on each CP rank.**
If the target is not split along the sequence dimension (i.e., unpack_latents(output) is applied), then the total loss value printed by framework becomes scaled by **1/CP**.

Also, during backpropagation, the loss gradient is scaled by a **CP factor** compared to non-CP training. Later, when gradients are accumulated across the cp_mesh via all_reduce, they are divided back by the same factor.
image

Therefore, torchtitan’s design implicitly requires both input and target to be split along seqlen when computing loss under CP.

File: torchtitan/experiments/flux/train.py (hunk 3 above)
Comment (limou102) [reply torchtitan/experiments/flux/train.py]: removed

File: torchtitan/experiments/flux/train.py (hunk 1 above)
Comment (limou102) [reply torchtitan/experiments/flux/train.py]: replaced it with self.parallel_dims.cp_enabled. 
I had previously used cp_group because of the implementation below that calls unpack_latents() before computing the loss, but I found that both the loss values and the gradients were incorrect — an explanation follows below.

**Comment by limou102** [comment]
**2025-10-14 update**
I conducted an additional experiment using global_batch_size=128 and iterations=20000, training two tasks:
Task A: local_batch=16, fsdp=8, cp=1
Task B: local_batch=32, fsdp=4, cp=2

The validation dataset loss curves for both tasks are nearly identical.
image


Moreover, when performing actual inference using the checkpoints from iterations=20000, the generated images both align well with the prompts.
image

Below is all test case outputs.
[inference_results_gb128_iter20000.zip](https://github.com/user-attachments/files/22895586/inference_results_gb128_iter20000.zip)

**Comment by fegin** [comment]
Can you also paste memory usage?

**Comment by limou102** [comment]
> Can you also paste memory usage?

image
The log above corresponds to local_batch=32, cp=2, fsdp=4, and the one below is for local_batch=16, cp=1, fsdp=8.
In this case, the sequence length is quite small, so enabling CP doesn’t show a significant benefit — instead, it increases memory usage by about 10GB (132GB vs 122GB).

File: torchtitan/experiments/flux/train.py [hunk 4]
```diff
@@ -132,24 +132,49 @@ def forward_backward_step(
...
-                img_ids=latent_pos_enc,
-                txt=t5_encodings,
-                txt_ids=text_pos_enc,
-                y=clip_encodings,
-                timesteps=timesteps,
-            )
-
-            # Convert sequence of patches to latent shape
-            pred = unpack_latents(latent_noise_pred, latent_height, latent_width)
+        with torch.no_grad(), torch.device(self.device):
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py]: Can we remove this line, and share the context with code above? Similarly can we remove the one above as well, and only keep one `with torch.no_grad(), torch.device(self.device):` in this function?

File: torchtitan/experiments/flux/train.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/train.py]: I get the idea. Thanks for explaining!

File: torchtitan/experiments/flux/train.py [hunk 5]
```diff
@@ -132,24 +132,49 @@ def forward_backward_step(
...
-                img=latents,
-                img_ids=latent_pos_enc,
-                txt=t5_encodings,
-                txt_ids=text_pos_enc,
-                y=clip_encodings,
-                timesteps=timesteps,
-            )
-
-            # Convert sequence of patches to latent shape
-            pred = unpack_latents(latent_noise_pred, latent_height, latent_width)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py:146]: please run linting before push
```
pre-commit run --all-files
```
see https://github.com/pytorch/torchtitan/blob/main/CONTRIBUTING.md#pull-requests

**Comment by limou102** [comment]
> Could you add an integration test here, with FSDP+CP? https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/tests/integration_tests.py
> 
> Could you help update supported features and TODO in README.md? https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/README.md#supported-features
> 
> In terms of loss comparison, what you did makes sense. A more recommended and rigorous approach is to make sure both data loading behavior and parameter init to be exactly the same across two runs. So instead of
> 
> > I conducted an additional experiment using global_batch_size=128 and iterations=20000, training two tasks:
> > Task A: local_batch=16, fsdp=8, cp=1
> > Task B: local_batch=32, fsdp=4, cp=2
> 
> We can do the following to make sure data loading behaviors align, which only depends on DP degree (batch sharding degree)
> 
> * Task A: local_batch=16, fsdp=2, cp=1 (2 GPU)
> * Task B: local_batch=16, fsdp=2, cp=4 (8 GPU)
> 
> Then we can use a seed checkpoint to make sure model parameters are the same, see https://github.com/pytorch/torchtitan/blob/main/docs/debugging.md#seed-checkpoint-based-reproducibility
> 
> Finally we can fix seed (training.seed) so that random behavior (e.g. dropout) is consistent, too.


I first generated a seed checkpoint, then ran 2 tasks with deterministic=True:
Task A: local_batch=16, fsdp=2, cp=1 using 2 GPUs
Task B: local_batch=16, fsdp=2, cp=4 using 8 GPUs

Their loss curves and values are completely identical.
image

image

In addition, integration test and updated to Flux README were added.

File: torchtitan/experiments/flux/infra/parallelize.py [hunk 6]
```diff
@@ -46,6 +46,16 @@ def parallelize_flux(
         else:
             logger.info("Applied FSDP to the model")
 
+        if parallel_dims.cp_enabled:
+            try:
+                from torch.distributed.tensor.experimental._attention import _cp_options
+
+                _cp_options.enable_load_balance = False
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/infra/parallelize.py]: why setting it to False here?

Comment (limou102) [reply torchtitan/experiments/flux/infra/parallelize.py]: In PyTorch, this is an experimental module where _cp_options.enable_load_balance = True is hardcoded by default.
https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_attention.py#L71

In this state, only attention computations with is_causal=True are supported.
However, for multimodal models like Flux, where is_causal=False in attention, you need to manually set _cp_options.enable_load_balance to False.
https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_attention.py#L389

Comment (limou102) [reply torchtitan/experiments/flux/infra/parallelize.py]: This module only affects ring_attention computation under Context Parallelism.

Comment (tianyu-l) [reply torchtitan/experiments/flux/infra/parallelize.py]: @limou102
Got it, but why do you do the try-catch here?

I think the place to set this is tricky. If you use the context manager in `train.py`, it'll anyway shard the input along seq dimension. If it's going to the else path, `_cp_options.enable_load_balance` would be `True` and I expect the program would always fail (because of this line https://github.com/pytorch/pytorch/blob/d0c24b392cbb7b213d22e42c52c6c2d1ac2da1bd/torch/distributed/tensor/experimental/_attention.py#L388-L389)

What stops us from always setting it to `False`?

cc @fegin as you are planning next steps. I think we can land this for now.

Comment (limou102) [reply torchtitan/experiments/flux/infra/parallelize.py]: The reason I used a try-catch is that this code is part of PyTorch’s experimental module, and it might change in the future — for example, _cp_options could be moved elsewhere.
If that happens, the catch block will trigger a log message to notify us early.
In theory, if the catch is triggered, we could call logger.error() and then abort.

As for why i don’t always set enable_load_balance to False, I’m not entirely sure whether that would cause a performance regression, however, setting it to False has been verified to be correct in ring_attention when is_causal=False.

Comment (lxgsbqylbk) [reply torchtitan/experiments/flux/infra/parallelize.py]: > @limou102 Got it, but why do you do the try-catch here?
> 
> I think the place to set this is tricky. If you use the context manager in `train.py`, it'll anyway shard the input along seq dimension. If it's going to the else path, `_cp_options.enable_load_balance` would be `True` and I expect the program would always fail (because of this line https://github.com/pytorch/pytorch/blob/d0c24b392cbb7b213d22e42c52c6c2d1ac2da1bd/torch/distributed/tensor/experimental/_attention.py#L388-L389)
> 
> What stops us from always setting it to `False`?
> 
> cc @fegin as you are planning next steps. I think we can land this for now.

_cp_options.enable_load_balance should be False if it's going to else path, since no exception occured

Comment (tianyu-l) [reply torchtitan/experiments/flux/infra/parallelize.py]: @lxgsbqylbk sorry I didn't mean to say "else path", I wanted to say "except path". In that case I believe the behavior is unexpected (we do input sharding, but not setting load_balance to False)

@limou102 
Given that you can't achieve correct behavior with try-catch, I'd suggest we just do `_cp_options.enable_load_balance = False` without try-catch for now. I believe @fegin will anyway do some refactor on this code soon.

Comment (limou102) [reply torchtitan/experiments/flux/infra/parallelize.py]: The try-catch block has been removed. This way, if the torch experimental module changes in future updates while Flux CP is enabled, it will raise an error directly, alerting the user immediately.

File: torchtitan/experiments/flux/infra/parallelize.py [hunk 7]
```diff
@@ -46,6 +46,12 @@ def parallelize_flux(
         else:
             logger.info("Applied FSDP to the model")
 
+        if parallel_dims.cp_enabled:
+            from torch.distributed.tensor.experimental._attention import _cp_options
+
+            _cp_options.enable_load_balance = False
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/infra/parallelize.py:56]: Please add a comment here explaining the attention in FLUX is not using causal masks, so CP doesn't support load balancing.

**Review by tianyu-l — COMMENTED** [review]
Looks good to me.

Could you please also help update the validator for FLUX https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/validate.py

You could refer to how it's done for the language model https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/validate.py#L114

File: torchtitan/experiments/flux/infra/parallelize.py (hunk 7 above)
Comment (limou102) [reply torchtitan/experiments/flux/infra/parallelize.py:56]: comment added

File: torchtitan/experiments/flux/infra/parallelize.py (hunk 6 above)
Comment (limou102) [reply torchtitan/experiments/flux/infra/parallelize.py]: Added Context Parallelism in the validation phase(however, since the main performance bottleneck typically lies in training, CP doesn’t seem particularly necessary here...)
The validation loss curve remains consistent with the previous results.

File: torchtitan/experiments/flux/infra/parallelize.py (hunk 7 above)
Comment (limou102) [reply torchtitan/experiments/flux/infra/parallelize.py:56]: Added Context Parallelism in the validation phase(however, since the main performance bottleneck typically lies in training, CP doesn’t seem particularly necessary here).
The validation loss curve remains consistent with the previous results.

File: torchtitan/experiments/flux/validate.py [hunk 8]
```diff
@@ -212,23 +212,46 @@ def validate(
...
+                            img=latents,
+                            img_ids=latent_pos_enc,
+                            txt=t5_encodings,
+                            txt_ids=text_pos_enc,
+                            y=clip_encodings,
+                            timesteps=timesteps,
+                        )
 
-            # Convert sequence of patches to latent shape
-            pred = unpack_latents(latent_noise_pred, latent_height, latent_width)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/validate.py:227]: Not removing `unpack_latents` from import will cause linter to complain.
